types/src/deploy.rs

```diff
@@ -0,0 +1,1780 @@
+
+        let mut deploy = Deploy {
+            hash,
+            header,
+            payment,
+            session,
+            approvals: BTreeSet::new(),
+            #[cfg(any(feature = "once_cell", test))]
+            is_valid: OnceCell::new(),
+        };
+
+        if let Some(secret_key) = account_and_secret_key.secret_key() {
+            deploy.sign(secret_key);
+        }
+        deploy
+    }
+
+    /// Returns the `DeployHash` identifying this `Deploy`.
+    pub fn hash(&self) -> &DeployHash {
+        &self.hash
+    }
+
+    /// Returns the public key of the account providing the context in which to run the `Deploy`.
+    pub fn account(&self) -> &PublicKey {
+        self.header.account()
+    }
+
+    /// Returns the creation timestamp of the `Deploy`.
+    pub fn timestamp(&self) -> Timestamp {
+        self.header.timestamp()
+    }
+
+    /// Returns the duration after the creation timestamp for which the `Deploy` will stay valid.
+    ///
+    /// After this duration has ended, the `Deploy` will be considered expired.
+    pub fn ttl(&self) -> TimeDiff {
+        self.header.ttl()
+    }
+
+    /// Returns `true` if the `Deploy` has expired.
+    pub fn expired(&self, current_instant: Timestamp) -> bool {
+        self.header.expired(current_instant)
+    }
+
+    /// Returns the price per gas unit for the `Deploy`.
+    pub fn gas_price(&self) -> u64 {
+        self.header.gas_price()
+    }
+
+    /// Returns the hash of the body (i.e. the Wasm code) of the `Deploy`.
+    pub fn body_hash(&self) -> &Digest {
+        self.header.body_hash()
+    }
+
+    /// Returns the name of the chain the `Deploy` should be executed on.
+    pub fn chain_name(&self) -> &str {
+        self.header.chain_name()
+    }
+
+    /// Returns a reference to the `DeployHeader` of this `Deploy`.
+    pub fn header(&self) -> &DeployHeader {
+        &self.header
+    }
+
+    /// Consumes `self`, returning the `DeployHeader` of this `Deploy`.
+    pub fn take_header(self) -> DeployHeader {
+        self.header
+    }
+
+    /// Returns the `ExecutableDeployItem` for payment code.
+    pub fn payment(&self) -> &ExecutableDeployItem {
+        &self.payment
+    }
+
+    /// Returns the `ExecutableDeployItem` for session code.
+    pub fn session(&self) -> &ExecutableDeployItem {
+        &self.session
+    }
+
+    /// Returns the `Approval`s for this deploy.
+    pub fn approvals(&self) -> &BTreeSet<Approval> {
+        &self.approvals
+    }
+
+    /// Adds a signature of this `Deploy`'s hash to its approvals.
+    pub fn sign(&mut self, secret_key: &SecretKey) {
+        let approval = Approval::create(&self.hash, secret_key);
+        self.approvals.insert(approval);
+    }
+
+    /// Returns the `ApprovalsHash` of this `Deploy`'s approvals.
+    pub fn compute_approvals_hash(&self) -> Result<ApprovalsHash, bytesrepr::Error> {
+        ApprovalsHash::compute(&self.approvals)
+    }
```

Comment:
does it make sense to add a OnceCell non-serialized field similar to is_valid to avoid computing this more than once?

Reply:
I'll do a ticket request to investigate.  If this is only called once or twice per deploy, it might not be worthwhile.

Reply:
That's sensible. :+1: